lib/upgrader.js

```diff
@@ -0,0 +1,95 @@
+module.exports = class Upgrader {
+  constructor (session, hooks = {}) {
+    this.session = session
+    this.active = null
+    this.preupgrade = hooks.preupgrade || null
+  }
+
+  cancel () {
+    if (this.active) {
+      this.active.cancel()
+      this.active = null
+    }
+  }
+
+  onupgrade (request) {
+    if (this.active) this.active.cancel()
+
+    const upgrade = this.active = new Upgrade(this)
+    upgrade.run(request)
+
+    return this.active.upgraded
+  }
+
+  ontruncate () {
+    // TODO: handle truncation
+  }
+}
+
+class Upgrade {
+  constructor (upgrader) {
+    this.upgrader = upgrader
+    this.active = true
+    this.upgraded = new Promise((resolve) => {
+      this.done = resolve
+    })
+  }
+
+  cancel () {
+    this.active = false
+  }
+
+  async run (request) {
+    const session = this.upgrader.session
+    const preupgradeLength = session.length
+
+    if (this.upgrader.preupgrade) session._updateSnapshot()
+
+    let upgraded = await request
```

Comment:
whats stopping us from awating the request in update itself? seems cleaner

Reply:
Need to do `session._updateSnapshot()` before the update is applied to the core to keep the session on the preupgrade length.

Reply:
Is session here, the session the user called .update on?

Reply:
Yep (`this`):

https://github.com/hypercore-protocol/hypercore-next/blob/94e4e3a7d133fb2386cbf5d9a69e8e039417a239/index.js#L83

Reply:
Ok, will see what we can do to improve this, but ok for now.